core/game.h

```diff
@@ -235,9 +235,17 @@ to look into this) if the strategy is identical to knuth’s.
       JNIEnv* jni_env = Ludii::JNIUtils::GetEnv();
 
       if (jni_env) {
-        Ludii::LudiiGameWrapper game_wrapper(jni_env, ludii_name);
-        state_ = std::make_unique<Ludii::LudiiStateWrapper>(
-            seed, jni_env, std::move(game_wrapper));
+        if (ludii_name == Ludii::LudiiGameWrapper::last_ludii_game_name) {
+          // We can recycle previously-constructed Java object
+          state_ = std::make_unique<Ludii::LudiiStateWrapper>(
+              seed,
+              std::move(Ludii::LudiiGameWrapper::last_ludii_game_wrapper));
```

Comment:
This caching of last_ludii_game_wrapper does not seem thread-safe to me. Am I missing something?

Reply:
This is the Java object, and on the Java side this has been built to be re-usable across multiple threads. This "game wrapper" in Ludii does not represent a single game in the sense of an "episode" or a "round". It's not "this instance of Chess we played last night". These objects represent the overall game as a whole, i.e. the rules that the game is played by, and those can be freely shared among many threads. Only the LudiiStateWrapper object is unique to the episode it's a part of (or really just it's a unique single game state), so that one should not be shared in any way.

Reply:
Ok, but it probably shouldn't be moved then? Moving implies that it can only be used by one state instance, and it can not be shared.

Also, the game name string comparison and assignment is not thread safe

Reply:
Ok yeah you're right, especially little things like the string comparison theoretically not being thread safe are annoying. Maybe it's easier to just get rid of this recycling again, rather than trying to implement a thread-safe solution? I did this part originally because profiling suggested that a lot of time was being spent here, but after implementing the recycling the profiling still said exactly the same... so probably the profiling wasn't accurate anyway. The idea is still theoretically nice, but the payoff in practice doesn't seem to be worth it...

Reply:
I'm just not sure that the recycling is working as you expect, which may be why it's not faster. Specifically, what happens when the game wrapper object is moved from?

Reply:
The LudiiStateWrapper constructor just creates another shared pointer to it. The main speedup that I was looking for is that it would avoid calling the Java-version of the LudiiGameWrapper constructor again, because that one can be relatively expensive. And that does indeed get avoided in this way. With "relatively expensive" I still mean like a couple of seconds though (maybe 10-20 seconds for really huge games, but less than a second for simple games), not minutes or hours.

But after I saw that this change had no effect in terms of profiling results, I also tried printing when all the constructors get called, and notice that they all really only just get called a few times right at the beginning of training anyway. Not super often either, probably around 10 times or so. Then during an hour-long profiling of the replay buffer warmup, none of these constructors ever got called again. So I really don't trust the profiling results that continued showing this stuff as a major time contributor anyway, and it seems to cause more problems than it's worth.